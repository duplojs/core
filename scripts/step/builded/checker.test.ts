import { Checker, CheckerStep, makeFloor, zod, type CheckerStepParams } from "../..";
import { Response } from "@scripts/response";
import { BuildedCheckerStep } from "./checker";
import { duploTest } from "@test/utils/duploTest";

describe("BuildedCheckerStep", () => {
	const checker = new Checker("test");
	checker.setOptions({
		toto: 1,
		test: "",
	});

	it("merge object options", () => {
		const params: CheckerStepParams = {
			input: () => ({}),
			catch: () => new Response(300, "test", 11),
			result: "",
			options: {
				toto: 2,
				test1: "&",
			},
		};

		const step = new CheckerStep(checker, params);

		const buildedCheckerStep = new BuildedCheckerStep(duploTest, step);

		expect(buildedCheckerStep.params.options).toStrictEqual({
			toto: 2,
			test: "",
			test1: "&",
		});

		expect(
			buildedCheckerStep.checkerFunction(
				1,
				(info, data) => ({
					info,
					data,
				}),
				{},
			),
		).toStrictEqual({
			info: "<([{|none|}])>",
			data: undefined,
		});
	});

	it("merge function options", () => {
		const params: CheckerStepParams = {
			input: () => ({}),
			catch: () => new Response(300, "test", 11),
			result: "",
			options: () => ({
				toto: 2,
				test1: "&",
			}),
		};

		const step = new CheckerStep(checker, params);

		const buildedCheckerStep = new BuildedCheckerStep(duploTest, step);

		expect(
			typeof buildedCheckerStep.params.options === "function"
				? buildedCheckerStep.params.options(makeFloor().pickup)
				: undefined,
		).toStrictEqual({
			toto: 2,
			test: "",
			test1: "&",
		});
	});

	it("toString options: function, result: string, indexing", async() => {
		const params: CheckerStepParams = {
			input: () => ({}),
			catch: () => new Response(300, "test", 11),
			result: "",
			indexing: "toto",
			options: () => ({
				toto: 2,
				test1: "&",
			}),
		};

		const step = new CheckerStep(checker, params, [new Response(100, "toto", zod.undefined())]);

		const buildedCheckerStep = new BuildedCheckerStep(duploTest, step);

		await expect(buildedCheckerStep.toString(1)).toMatchFileSnapshot("__data__/checker1.txt");
	});

	it("toString options: object, result: array", async() => {
		const params: CheckerStepParams = {
			input: () => ({}),
			catch: () => new Response(300, "test", 11),
			result: ["", "tot"],
			options: {
				toto: 2,
				test1: "&",
			},
			skip: () => true,
		};

		const step = new CheckerStep(checker, params, [new Response(100, "toto", zod.undefined())]);

		const buildedCheckerStep = new BuildedCheckerStep(duploTest, step);

		await expect(buildedCheckerStep.toString(1)).toMatchFileSnapshot("__data__/checker2.txt");
	});
});
