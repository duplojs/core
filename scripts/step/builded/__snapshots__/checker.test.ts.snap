// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BuildedCheckerStep > toString options: function, result: string, indexing 1`] = `
"
			
/* step-checker-(1)-before */
/* end_block */


			result = this.steps[1].checkerFunction(
				this.steps[1].params.input(floor.pickup),
				(info, data) => ({info, data}),
				this.steps[1].params.options(floor.pickup),
			);

			
/* step-checker-(1)-before-treat-result */
/* end_block */


			if(this.steps[1].params.result !== result.info){
				result = this.steps[1].params.catch(
					result.info, 
					result.data, 
					floor.pickup
				);

				
				let temp = this.steps[1].responseZodSchema.safeParse(result);

				if(!temp.success){
					throw new this.ContractResponseError(temp.error, result);
				}
			

				break Execution;
			}

			
/* step-checker-(1)-before-indexing */
/* end_block */


			floor.drop(this.steps[1].params.indexing, result.data)

			
/* step-checker-(1)-after */
/* end_block */

			"
`;

exports[`BuildedCheckerStep > toString options: object, result: array 1`] = `
"
		
/* step-skip-(1)-before */
/* end_block */


		if(!this.steps[1].params.skip(floor.pickup)){
			
			
/* step-checker-(1)-before */
/* end_block */


			result = this.steps[1].checkerFunction(
				this.steps[1].params.input(floor.pickup),
				(info, data) => ({info, data}),
				this.steps[1].params.options,
			);

			
/* step-checker-(1)-before-treat-result */
/* end_block */


			if(!this.steps[1].params.result.includes(result.info)){
				result = this.steps[1].params.catch(
					result.info, 
					result.data, 
					floor.pickup
				);

				
				let temp = this.steps[1].responseZodSchema.safeParse(result);

				if(!temp.success){
					throw new this.ContractResponseError(temp.error, result);
				}
			

				break Execution;
			}

			
/* step-checker-(1)-before-indexing */
/* end_block */


			

			
/* step-checker-(1)-after */
/* end_block */

			
		}
		"
`;
